foundryup/src/process.rs

```diff
@@ -0,0 +1,342 @@
+use std::{
+    any::{Any, TypeId},
+    cell::RefCell,
+    env,
+    ffi::OsString,
+    fmt, io,
+    ops::Deref,
+    path::PathBuf,
+    ptr::NonNull,
+    sync::{
+        atomic::{AtomicBool, AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+thread_local! {
+    static CURRENT_STATE: State = State {
+        current: RefCell::new(Process::default()),
+    };
+}
+
+static EXISTS: AtomicBool = AtomicBool::new(false);
+static CURRENT_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+// tracks the state of `GLOBAL_PROCESS`
+static GLOBAL_PROCESS_STATE: AtomicUsize = AtomicUsize::new(UN_SET);
+
+const UN_SET: usize = 0;
+const SETTING: usize = 1;
+const SET: usize = 2;
+
+static mut GLOBAL_PROCESS: Option<Process> = None;
```

Comment:
can you remind me where this pattern comes from and why it's needed?

Reply:
this is now a mix of the global+local pattern we use in the solc reporter which is based on the tracing dispatcher and the rustup threadlocal process.

perhaps we don't need a global instance and instead only use thread local, but having access to shared instance makes it easy to get access to it from anywhere in the code while controlling the type that's accessed. and makes it easier to swap out the real `Process` for a `TesProcess` for testing